### PR TITLE
Wip/hughsie/review metadata

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: libfwupdplugin/fu-device.c [hunk 1]
```diff
@@ -2972,6 +2972,63 @@ fu_device_probe_invalidate (FuDevice *self)
...
+	FuDeviceClass *klass = FU_DEVICE_GET_CLASS (self);
+	g_autoptr(GHashTable) metadata = NULL;
+
+	g_return_val_if_fail (FU_IS_DEVICE (self), NULL);
+
+	/* metadata for all devices */
+	metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+	g_hash_table_insert (metadata,
+			     g_strdup ("GType"),
+			     g_strdup (G_OBJECT_TYPE_NAME (self)));
```
Comment (superm1) [inline libfwupdplugin/fu-device.c:2997]: Is this actually valuable for all the devices?

Comment (hughsie) [reply libfwupdplugin/fu-device.c:2997]: For things like VLI or Wacom that are subclassed quite deeply, yes. For others perhaps less so.

File: libfwupdplugin/fu-device.c [hunk 2]
```diff
@@ -2972,6 +2972,63 @@ fu_device_probe_invalidate (FuDevice *self)
...
+
+	g_return_val_if_fail (FU_IS_DEVICE (self), NULL);
+
+	/* metadata for all devices */
+	metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+	/* optionally subclassed */
+	if (klass->report_metadata_post != NULL)
+		klass->report_metadata_post (self, metadata);
+	return g_steal_pointer (&metadata);
```
Comment (superm1) [inline libfwupdplugin/fu-device.c:3029]: Most device types won't have this new vfunc, I wonder if it's really worth the extra memory to create the hash table and return an empty hash table in that common case.  Instead why not

```
if (klass->report_metadata_post == NULL)
        return NULL;
metadata = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
klass->report_metadata_post (self, metadata);
return g_steal_pointer (&metadata);
```

File: libfwupdplugin/fu-device.c (hunk 1 above)
Comment (superm1) [reply libfwupdplugin/fu-device.c:2997]: But so in those it actually matters, shouldn't they be device metadata made by the plugin?

File: libfwupdplugin/fu-device.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-device.c:3029]: Sure, but this isn't fast path at all, so I went for clarity. It's called just once at startup if it's the first boot after a UEFI update.

File: libfwupdplugin/fu-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-device.c:2997]: Sure, I can do that if you think it's clearer.

Comment (superm1) [reply libfwupdplugin/fu-device.c:2997]: Yes I do think it's clearer.

File: libfwupdplugin/fu-device.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-device.c:3029]: I guess if `GType` isn't always added then my arguement is null and void. Will fix.

File: libfwupdplugin/fu-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-device.c:2997]: Okay, fixed.

**Comment by hughsie** [comment]
> to adjust too to be more device centric (tpm-eventlog

That one is a bit special as we add it as a dep to the UEFI devices.

> dell-dock

Would this need a vfunc implemented for all subclassed FuDevices? I don't know if it's just needed for EC etc.

**Comment by superm1** [comment]
> Would this need a vfunc implemented for all subclassed FuDevices? I don't know if it's just needed for EC etc.

It's relevant information for all of them yeah.  So maybe the way it's done now is fine.

**Pull request merged**